src/java.base/share/classes/sun/util/locale/LanguageTag.java

```diff
@@ -414,6 +407,50 @@ private boolean parsePrivateuse(StringTokenIterator itr, ParseStatus sts) {
         return found;
     }
 
+    public static String caseFoldTag(String tag) {
+        LanguageTag parsedTag = parse(tag, null);
+        // Illegal tags
+        if (parsedTag.parseSts.errorMsg != null) {
+            throw new IllformedLocaleException(String.format("Ill formed tag:" +
+                    " %s", parsedTag.parseSts.errorMsg));
+        }
+        // Legacy tags
+        String potentialLegacy = tag.toLowerCase(Locale.ROOT);
+        if (LEGACY.containsKey(potentialLegacy)) {
+            return LEGACY.get(potentialLegacy)[0];
+        }
+        // Non-legacy tags
+        StringBuilder bldr = new StringBuilder();
+        String[] subtags = tag.split("-");
+        boolean privateFound = false;
+        boolean singletonFound = false;
+        boolean privUseVarFound = false;
+        for (int i = 0; i < subtags.length; i++) {
+            if (privUseVarFound) {
```

Comment:
The following block of code might be easier to read if subtags[i] was put in a local.